### PR TITLE
feat: replace sequence and confirmBy with hostBlockNumber

### PR DESCRIPTION
**Pull request by anna-carroll**
[Linear ticket](https://linear.app/initiates/issue/ENG-262/remove-sequence-and-confirmby-add-hostblocknumber)

## Motivation 

The node will mine one RU block per host block. 

## Solution

As such, we can remove the `sequence` number (that tracks the rollup block number and ensures blocks are mined in correct order) and the `confirmBy` field (that ensures blocks are mined by a certain time). 

We can replace these fields with a `hostBlockNumber` field (that ensures that the rollup block is mined in the exact specified host block, encapsulating block ordering, block numbers, and expiration times within one field).

As a convenience, we will also store the `deployBlockNumber` - the host block number at which the contract was deployed. Rollup block numbers can be derived from this value.

### CODE REVIEW

File: src/Zenith.sol [hunk 1]
```diff
@@ -129,19 +100,14 @@ contract Zenith is Passage {
...
     /// @custom:emits BlockSubmitted if the block is successfully submitted.
     function submitBlock(BlockHeader memory header, uint8 v, bytes32 r, bytes32 s, bytes calldata) external {
-        // assert that the sequence number is valid and increment it
-        uint256 _nextSequence = incrementSequence(header.rollupChainId);
-        if (_nextSequence != header.sequence) revert BadSequence(_nextSequence);
-
-        // assert that confirmBy time has not passed
-        if (block.timestamp > header.confirmBy) revert BlockExpired();
+        // assert that the host block number is valid
+        if (block.number != header.hostBlockNumber) revert IncorrectHostBlock();
```
Comment (anna-carroll) [inline src/Zenith.sol:110]: note @prestwich : i realized while implementing that this line is going to mess up default gas estimation for submitting blocks. during gas estimation, `header.hostBlockNumber` should be `current block.number + 1`  (e.g. the next block that will mine). 

i think Builders would need to override checks that the transaction will fail before submitting, which is not ideal

File: .gas-snapshot [hunk 2]
```diff
@@ -1,6 +1,5 @@
...
-ZenithTest:test_blockExpired() (gas: 55499)
-ZenithTest:test_notSequencer() (gas: 58618)
-ZenithTest:test_onePerBlock() (gas: 104426)
-ZenithTest:test_submitBlock() (gas: 88526)
\ No newline at end of file
+ZenithTest:test_badSignature() (gas: 37377)
+ZenithTest:test_incorrectHostBlock() (gas: 35565)
+ZenithTest:test_notSequencer() (gas: 34145)
+ZenithTest:test_onePerBlock() (gas: 68700)
+ZenithTest:test_submitBlock() (gas: 61119)
```
Comment (anna-carroll) [inline .gas-snapshot:5]: very nice reduction in gas costs

File: src/Zenith.sol (hunk 1 above)
Comment (prestwich) [reply src/Zenith.sol:110]: within `eth_call` and `eth_estimateGas`, `block.number` is the next (unmined) block number unless you tell it otherwise

Comment (prestwich) [reply src/Zenith.sol:110]: you can see this behavior here:
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-eth-api/src/helpers/pending_block.rs#L79-L82

Comment (anna-carroll) [reply src/Zenith.sol:110]: slay

File: src/Zenith.sol [hunk 3]
```diff
@@ -6,30 +6,25 @@ import {Passage} from "./Passage.sol";
 contract Zenith is Passage {
     /// @notice The address that is allowed to set/remove sequencers.
     address public immutable sequencerAdmin;
+    /// @notice The block number at which the Zenith contract was deployed.
+    uint256 public immutable deployBlockNumber;
```
Comment (prestwich) [inline src/Zenith.sol:10]: if it's an immutable it gets stored in a code region, reading would be slightly easier+cheaper if this were stored in state instead

File: test/Zenith.t.sol [hunk 4]
```diff
@@ -41,29 +38,16 @@ contract ZenithTest is Test {
...
-    function test_blockExpired() public {
-        // change to expired confirmBy time
-        header.confirmBy = block.timestamp - 1;
-        commit = target.blockCommitment(header);
-
-        // sign block commitmenet with sequencer key
-        (uint8 v, bytes32 r, bytes32 s) = vm.sign(sequencerKey, commit);
-
-        vm.expectRevert(abi.encodeWithSelector(Zenith.BlockExpired.selector));
+        vm.expectRevert(abi.encodeWithSelector(Zenith.IncorrectHostBlock.selector));
```
Comment (prestwich) [inline test/Zenith.t.sol:50]: probably don't need the encode step?

Comment (anna-carroll) [reply test/Zenith.t.sol]: ah correct, fixed: https://github.com/init4tech/zenith/pull/48/commits/d7cf45865d3457eedb41e6c2c86f81c01743acc5

File: src/Zenith.sol (hunk 3 above)
Comment (anna-carroll) [reply src/Zenith.sol]: gotcha, was assuming it'd be read through RPC interface for some reason? changed: https://github.com/init4tech/zenith/pull/48/commits/d462e0905ca7dd9c5fd8aa97ff0092961d548576

**Review by prestwich — APPROVED** [review]

**Pull request merged**